File: ADT/segment_tree/range-min-query.cpp
//Using Segment Tree
//T.C : O(q*log(n))
//S.C : O(4*n)

void buildSegmentTree(int i, int l, int r, int segmentTree[], int arr[]){
    if(l == r){
        segmentTree[i] = arr[l];
        return;
    }
    int mid = l + (r - l) / 2;
    buildSegmentTree(2 * i + 1, l, mid, segmentTree, arr);
    buildSegmentTree(2 * i + 2, mid + 1, r, segmentTree, arr);
    segmentTree[i] = min(segmentTree[2 * i + 1], segmentTree[2*i + 2]);
}
int *constructST(int arr[], int n) {
    int* segmentTree = new int[4 * n];
    buildSegmentTree(0, 0, n - 1, segmentTree, arr);
    return segmentTree;
}

int query(int start, int end, int i, int l, int r, int segmentTree[]){
    if(l > end || r < start){
        return INT_MAX;
    }
    if(l >= start && r <= end){
        return segmentTree[i];
    }
    int mid = l + (r - l) / 2;
    return min(
        query(start, end, 2 * i + 1, l, mid, segmentTree),
        query(start, end, 2 * i + 2, mid + 1, r, segmentTree)
        );
}

int RMQ(int st[], int n, int a, int b) {
    return query(a, b, 0, 0, n-1, st);
}

File: ADT/trees/segment_tree/range_min_query.py
import math
from typing import List

def build_segment_tree(i: int, l: int, r: int, segment_tree: List[int], arr: List[int]) -> None:
    if l == r:
        segment_tree[i] = arr[l]
        return
    mid = l + (r - l) // 2
    build_segment_tree(2 * i + 1, l, mid, segment_tree, arr)
    build_segment_tree(2 * i + 2, mid+1,r, segment_tree, arr)
    segment_tree[i] = min(segment_tree[2*i + 1], segment_tree[2*i + 2])
    
def construct_st(arr: List[int]) -> List[int]:
    n = len(arr)
    segment_tree = [math.inf] * (4 * n)
    build_segment_tree(0, 0, n-1, segment_tree, arr)
    return segment_tree

def query(start: int, end: int, i: int, l: int, r: int, segment_tree: List[int]) -> int:
    if l > end or r < start:
        return math.inf
    if l >= start and r <= end:
        return segment_tree[i]
    mid = l + (r - l) // 2
    return min(
        query(start, end , 2 * i + 1, l, mid, segment_tree),
        query(start, end , 2 * i + 2, mid+1, r, segment_tree)
    )

def rmq(st: List[int], n: int, a: int, b: int) -> int:
    return query(a, b, 0, 0, n - 1, st)
